02.calendar/cal.js

```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env node
+
+const main = () => {
+  const options = parseArgs(process.argv.slice(2))
+  console.log(`      ${options.m}月 ${options.y}`)
+  const date = new Date(Date.UTC(options.y, --options.m))
+  process.stdout.write('日 月 火 水 木 金 土\n')
+  printCalendar(date)
+}
```

Comment:
どちらの方が読みやすいコードでしょうか?
変更後は前処理を前半にまとめて、後半は表示に専念する形

> console.log(`      ${options.m}月 ${options.y}`)

ただし上記のコードへの影響を無くすため、Dateクラスの月の指定で`-1`する作業が`--options.y`から`options. - 1`と変更する必要があり、微妙な感じになります。(コードを読むときに、`--`や`++`を利用した方が、0オリジン,1 オリジン調整だと推測しやすいと感じる。)

```suggestion
const main = () => {
  const options = parseArgs(process.argv.slice(2))
  const date = new Date(Date.UTC(options.y, options.m - 1))
  
  console.log(`      ${options.m}月 ${options.y}`)
  process.stdout.write('日 月 火 水 木 金 土\n')
  printCalendar(date)
}
```

Reply:
`--options.m`を使うと`m`の値が変わっちゃうのであまりよろしくないですね。
たぶんこんな動きになるんじゃないでしょうか？

```javascript
consolo.log(options.m) //=> 1
const date = new Date(Date.UTC(options.y, --options.m))
consolo.log(options.m) //=> 0
```

`options.m`の値はプログラムの実行中、一貫して同じ値になっている方が不具合や読み手の混乱を避けやすいので、その意味で`options.m - 1`のスタイルを推奨します。
なので後者のコードの方がベターかもしれません。

その一方で「変数の寿命を短くする」という観点で後者もまだ改善の余地があるかも？とは思います。

参考 [変数の寿命はできるだけ短くしよう \| FJORD BOOT CAMP（フィヨルドブートキャンプ）](https://bootcamp.fjord.jp/pages/311)

Reply:
> --options.mを使うとmの値が変わっちゃうのであまりよろしくないですね。

「ここで使用するのは最後だし問題ないだろう」という、プログラムを拡張するときに問題になる考え方で実装していました。指摘ありがとうございます。

---
>その一方で「変数の寿命を短くする」という観点で後者もまだ改善の余地があるかも？とは思います。

`options`と`date`という二つの変数で、年月を管理していたことがと考え、`options`の寿命を短くし、表示の処理では、`date`だけを利用するという方法を取るように変更しました。

```javascript
const main = () => {
  const options = parseArgs(process.argv.slice(2))
  const date = new Date(Date.UTC(options.y, options.m - 1))

  console.log(`      ${date.getMonth() + 1}月 ${date.getFullYear()}`)
  process.stdout.write('日 月 火 水 木 金 土\n')
  printCalendar(date)
}
```

---
副作用として以下の問題が発生中です。(対応方法を30分ほど考えても出てこないので意見をください。お願いします)
### 問題
入力(1オリジン)→Date(0オリジン)→表示(1オリジン)と
変換のたびのマジックナンバー(`+1`, `-1`)が発生している。
1. `const date = new Date(Date.UTC(options.y, options.m - 1))`:入力(1オリジン)からDate(0オリジン)へ
2. ``console.log(`     ${date.getMonth() + 1}月 ${date.getFullYear()}`)``:Date(0オリジン)を1オリジンで表示
3. `m: today.getMonth() + 1,`:入力のデフォルト値、Date(0オリジン)で取得して1オリジンで保存

<br>

↓もともと二つの変数を入り混じりで利用していたのは、`2.`の箇所で`+1`をしないで済ますためでした。

>`options`と`date`という二つの変数で、年月を管理していたことがと考え、`options`の寿命を短くし、表示の処理では、`date`だけを利用するという方法を取るように変更しました。

### 考えたこと
- JavaScriptのDateクラスの仕様上しかたないので放置
- もとの年月に関して2変数管理に戻す(変数の寿命は伸びるが、0オリジン用と1オリジン用で二つもっていたほうが効率的?)
- `toZeroOrigin()`, `toOneOrigin()`のようなメソッドを作る(逆にコードの可読性が下りそう)
- `const ADJUSTMENT_ORIGIN_NUM = 1`のようなマジックナンバーを置き変える定数を作成する。(オリジン調整用の定数の命名が難しい)

Reply:
そうですねー、僕だったらこうしますかね。

```javascript
  const options = parseArgs(process.argv.slice(2))
  console.log(`      ${options.m}月 ${options.y}`)
  console.log('日 月 火 水 木 金 土')

  const date = new Date(Date.UTC(options.y, options.m - 1, 1))
  printCalendar(date)
```

+1 や -1がある程度発生するのは仕方がないです。
人間の世界とJSの世界で数え方が違うので、その境界部分ではどうしても+1/-1が発生してしまいます。
もう少し具体的に言うとこんな感じですね。

- 起動時引数の指定 = 人間が入力するので1オリジン
- ターミナルに表示する「月」 = 人間が目にする値なので1オリジン
- new Dateの引数 = JSの世界なので0オリジン

なので、こういったタイミングで +1/-1 が発生するのはやむを得ないものとして許容する、でいいと思います。

> const ADJUSTMENT_ORIGIN_NUM = 1のようなマジックナンバーを置き変える定数を作成する。(オリジン調整用の定数の命名が難しい)

Dateの仕様はJSプログラマの基礎知識（一般常識？）と考えて差し支えないので、わざわざ定数を用意しなくても、+1/-1のままで大半のプログラマは理解可能だと思いますー。

Reply:
> なので、こういったタイミングで +1/-1 が発生するのはやむを得ないものとして許容する、でいいと思います。

かしこまりました。 
提案されたこちらの形式に変更致しました。
```javascript
  const options = parseArgs(process.argv.slice(2))
  console.log(`      ${options.m}月 ${options.y}`)
  console.log('日 月 火 水 木 金 土')

  const date = new Date(Date.UTC(options.y, options.m - 1, 1))
  printCalendar(date)
```